tests/unit/withOnyxTest.js

```diff
@@ -522,4 +523,98 @@ describe('withOnyxTest', () => {
     });
+
+    it('should render immediately when a onyx component is mounted a 2nd time', () => {
+        const TestComponentWithOnyx = withOnyx({
+            text: {
+                key: ONYX_KEYS.TEST_KEY,
+            },
+        })(ViewWithText);
+        const onRender = jest.fn();
+        let renderResult;
+
+        // Set the value in storage, but not in cache.
+        return StorageMock.setItem(ONYX_KEYS.TEST_KEY, 'test1')
+            .then(() => {
+                renderResult = render(<TestComponentWithOnyx onRender={onRender} />);
+
+                // The component should not render initially since we have no data in cache.
+                expect(onRender).not.toHaveBeenCalled();
+                return waitForPromisesToResolve();
+            })
+            .then(waitForPromisesToResolve)
+            .then(() => {
+                let textComponent = renderResult.getByText('test1');
+                expect(textComponent).not.toBeNull();
+                expect(onRender).toHaveBeenCalledTimes(1);
+
+                // Unmount the component and mount it again. It should now render immediately.
+                renderResult.unmount();
+                renderResult = render(<TestComponentWithOnyx onRender={onRender} />);
+
+                textComponent = renderResult.getByText('test1');
+                expect(textComponent).not.toBeNull();
+                expect(onRender).toHaveBeenCalledTimes(2);
+            });
+    });
+
+    it('should cache missing keys', () => {
+        const TestComponentWithOnyx = withOnyx({
+            text: {
+                key: ONYX_KEYS.TEST_KEY,
+            },
+        })(ViewWithText);
+        const onRender = jest.fn();
+
+        // Render with a key that doesn't exist in cache or storage.
+        let renderResult = render(<TestComponentWithOnyx onRender={onRender} />);
+
+        // The component should not render initially since we have no data in cache.
+        expect(onRender).not.toHaveBeenCalled();
+        return waitForPromisesToResolve().then(() => {
+            let textComponent = renderResult.getByText('null');
+            expect(textComponent).not.toBeNull();
+            expect(onRender).toHaveBeenCalledTimes(1);
+
+            // Unmount the component and mount it again. It should now render immediately.
+            renderResult.unmount();
+            renderResult = render(<TestComponentWithOnyx onRender={onRender} />);
+            textComponent = renderResult.getByText('null');
+            expect(textComponent).not.toBeNull();
+            expect(onRender).toHaveBeenCalledTimes(2);
+        });
+    });
+
+    it('should cache empty collections', () => {
+        const TestComponentWithOnyx = withOnyx({
+            text: {
+                key: ONYX_KEYS.COLLECTION.TEST_KEY,
+            },
+        })(ViewWithCollections);
+        const onRender = jest.fn();
+        let renderResult;
+
+        return StorageMock.setItem(ONYX_KEYS.SIMPLE_KEY, 'simple')
```

Comment:
Curious why do we need to add this key? is it just so there is something in the onyx storage?

Reply:
Yea to know if all keys are in cache we currently check if the array length is > 0 so if the storage is empty the check never works.